explain/explain.py

```diff
@@ -0,0 +1,753 @@
+
+    assert wrapped.__doc__
+    wrapped.__doc__ += """
+        Additional parameters:
+        hypothesis -- describe the hypothesis you are currently investigating.
+    """
+
+    return wrapped
+
+
+def revert_time_on_failure(
+    fn: Callable[Concatenate[UdbMcpGatewayAlias, P], T],
+) -> Callable[Concatenate[UdbMcpGatewayAlias, P], T]:
+    """
+    Decorator to ensure a tool does not change debugger time if an exception is thrown.
+    """
+
+    @functools.wraps(fn)
+    def wrapped(self, *args, **kwargs):
+        t = self.udb.time.get_bookmarked()
+        try:
+            return fn(self, *args, **kwargs)
+        except:
+            # Revert time on failure so failed tool invocations don't affect state.
+            self.udb.time.goto(t)
+            raise
+
+    return wrapped
+
+
+class UdbMcpGateway:
+    """
+    Plumbing class to expose selected UDB functionality as a set of tools to an MCP server.
+
+    The operations exposed are chosen to be well-suited to use by an LLM. For now, that means that
+    only reverse operations are supported (to impose reverse debugging practices on the LLM) and
+    that unnecessary (or potentially distracting) operations are not exposed.
+    """
+
+    def __init__(self, udb: udb_base.Udb):
+        self.udb = udb
+        self.mcp = FastMCP("UDB_Server", instructions=MCP_INSTRUCTIONS, log_level=LOG_LEVEL)
+        self.tools: list[str] = []
+        self._register_tools()
+
+    def _register_tools(self) -> None:
+        for name, fn in inspect.getmembers(self, inspect.ismethod):
+            if not name.startswith("tool_"):
+                continue
+            name = name.removeprefix("tool_")
+            self.tools.append(name)
+            self.mcp.add_tool(fn=fn, name=name, description=fn.__doc__)
+
+    @report
+    @source_context
+    @collect_output
+    @chain_of_thought
+    def tool_ugo_end(self) -> None:
+        """
+        Go to the end of recorded history.
+
+        The end of recorded history is a good place from which to begin investigating a bug.
+
+        If you get lost or see inconsistent state then return to the end and begin investigations
+        again.
+        """
+        self.udb.time.goto_end()
+
+    @report
+    @source_context
+    @collect_output
+    @chain_of_thought
+    def tool_last_value(self, expression: str) -> None:
+        """
+        Wind back time to the last time an expression was modified.
+
+        This should NOT be used with function call expressions e.g. my_function(number) as
+        it will be very slow.
+
+        This should NOT be used for address-taken expressions, such as &variable_name, since this
+        will be very slow and won't give a meaningful answer.
+
+        This should NOT be used for debugger convenience variables (starting with a $), since this
+        will be very slow.
+
+        Use expressions that are based solely on variables or memory locations.
```

Comment:
Does this work reliably? We could also use `gdb.parse_and_eval` and check that the restrictions are matched.